### ISSUE TITLE
Reference types vs value types refactoring

### ISSUE THREAD

**Issue by rfdickerson** [IBM-Swift/Kitura#355]
This ticket can be used for the refactoring of the data structures used in Kitura. Almost all of the data structures in Kitura are reference types (classes). Value-types offer many advantages such as:

- Thread-safety, since references do not need to be locked on mutation.
- No need for ARC, so possibility of memory cycles diminished.
- Prevents users from turning to inheritance.
- Compiler optimizations- allocation of memory cheap, usage with Copy on Write very fast.

[Building Better Apps with Value Types in Swift](https://developer.apple.com/videos/play/wwdc2015/414/)

Determining which to use is somewhat complex issue. Hopefully with the ticket we can create some sort of decision-making process for determining which kind of type to use.

## Summary

- Router

Keep as a **class**, the exact same instance should be shared by modules.

- RouterElement

Change to **struct**. This would allow us to populate the Router.routeElems with structs instead of references which would use a Swift Array vs. NSArray. Safety and performance optimizations possible.

- RouterMiddlewareGenerator

Keep as class?

- RouterRequest

Change to **struct**. 

- RouterResponse

Change to **struct**

- ContentType

Keep as class

- StaticFileServer

Keep as class ?

- BodyParser

Keep as class?

**Comment by shmuelk** [comment]
If RouterRequest and RouterResponse are structs how are middleware expected to do anything to a request or response as its being handled by the "pipeline"?

**Comment by rfdickerson** [comment]
The idea is that each stage of the "pipeline" structs go in- and structs go out. Given "copy on write" semantics, if changes need to be made (to a Response), the struct gets copied. If no changes are needed to be made the struct will pass through without modification.

**Comment by shmuelk** [comment]
Then your saying we will use inout. Because otherwise structs are pass by value and the copy on change happens only in the called function. Unless you want handlers and middleware to "return" results of some sort or other.

**Comment by rfdickerson** [comment]
No, not "inout"- rather return "new" structs after the handle function has processed them.

**Comment by shmuelk** [comment]
You mean return as in:

   func handle(request: RouterRequest, response: RouterResponse) -> RouterResponse

**Comment by rfdickerson** [comment]
Yeah, that's what I mean. Somewhat elegant way of handling Requests don't you think?

**Comment by shmuelk** [comment]
Actually, I think that it just seems so weird. Making me return my input parameters, just because they may have been changed.

**Comment by rfdickerson** [comment]
But it's composable, idempotent, and has no side-effects.

**Comment by shmuelk** [comment]
Copied after each time a middleware or handler does something.....

I somewhat disagree about the side affects, the RouterResponse that the Router gets at the end of the pipeline to work with is the one created somewhere along the way in the pipeline, not the one it created when things started.

**Comment by shmuelk** [comment]
Earlier in the thread you stated that using structs will prevent subclassing. I think that if the API is to return a request and response pair, then the API goes down the road of subclasses to wrap functionality.

You are working on issue #10. In the world being described in this issue, one would naturally wrap requests and responses with subclasses/substructs that add such capabilities.

Alternatively one could add  "lifecycle" calls to the request and response that do desired things at the right time.

Extensions is NOT the way to go here, because you often need to orchestrate/co-ordinate a set of middleware (i.e. Who is one first, what is on second.....)

**Comment by rfdickerson** [comment]
Right, I see what you mean regarding having Response object knowing that it's a CompressedResponse object. I think that's a tricky problem.

Inheritance is one way of solving the problem. But I recall some of the problems Dave Abraham brought up in his talk on [protocol-oriented programming](https://developer.apple.com/videos/play/wwdc2015/408/) about inheritance. Intuitively, I think that a Request and a Response remains a request and response, and subclassing is not needed. It should probably be the responsibility of someone else to ensure that the middleware are ordered properly in the array such that it makes sense. Compress, Encrypt, Send, etc.

**Comment by vadimeisenbergibm** [comment]
According to [this great post about structs in Swift](http://faq.sealedabstract.com/structs_or_classes/#swift-value-types), no class mentioned in this issue should be a struct. The author even gives HTTP Response as an example:
> What does it even mean to say that two HTTP responses are equal? (Pause for existential reflection.) That they have the same URL? They have the same contents? That one of them was once .copy()ed from another one (e.g., they have the same provenance)? I'm not sure that equality has a clear definition here.

tl;dr based on [Apple's Swift Programming Language Book](https://developer.apple.com/library/mac/documentation/Swift/Conceptual/Swift_Programming_Language/ClassesAndStructures.html/navigation/):
> The structure’s primary purpose is to encapsulate a few relatively simple data values.
...
In all other cases, define a class... In practice, this means that most custom data constructs should be classes, not structures.

My take on it - if you have a doubt should your type be a struct or a class, it should be a class.

**Comment by rfdickerson** [comment]
Ray says:

> If you’re not quite sure which mechanism applies to your situation, default to value types. You can always move to a class later with little effort.

https://www.raywenderlich.com/112027/reference-value-types-in-swift-part-1

This is exactly why all value types need to conform to the Equatable protocol: http://nshipster.com/swift-comparison-protocols/

**Comment by rfdickerson** [comment]
I would like to present this question, given that ~100% of our data structures are classes in Kitura. Are there any that could be migrated to be a value type instead?

**Comment by vadimeisenbergibm** [comment]
@rfdickerson Please see [my comment at raywenderlich](https://forums.raywenderlich.com/t/reference-vs-value-types-in-swift-part-1-2/6295/5?u=vadime)

My take on Apple guidelines is that value types should only be used for _relatively simple data values_ (as Apple put it). I am not aware of such types in Kitura, except for the enums we use.

**Comment by rfdickerson** [comment]
Let's consider this example:

**A:**
```swift
struct Point {
    let x: Float
    let y: Float
}

func scale(p: Point) -> Point {
....
}
```

**B:**
```swift
class Point {
    var x: Float
    var y: Float

    func scale() {
     ///
    }
}
````

**Comment by vadimeisenbergibm** [comment]
@rfdickerson 
1. I agree that Point is the best example for a struct.
2. Isn't using extension orthogonal to the issue? You can use extension for a class as well
3. Where is this statement from? What does it mean _default_ extensions?
_Objects should only be concerned with encapsulating values inside of them, one should leave behavior to default behavior extensions._
4. In general, it seems your comments are in favor of Functional Programming. Please note that Swift supports both paradigms - OOP and Functional. My belief is that neither paradigm is better - different paradigms are better for different problems. So we should decide on a problem-per-problem basis, which paradigm should we use.
5. According to the WWDC talk about value types, you have to perform copy on write on the reference types the value type is composed of, and as far as I understand, in a recursive way. This means that for RequestServer to be a value type, you have to copy on write SocketRequest and Socket with its Buffer. Seems not good for performance. There is also [a claim that equality is not easy to define for HTTP Response](http://faq.sealedabstract.com/structs_or_classes/#swift-value-types), the same claim valid for request as well. HTTP Request/Response do not look to me as simple types, with clearly defined equality semantics.

**Comment by rfdickerson** [comment]
I am hoping for a design of ServerRequest and ServerResponse to match something like this:

- [Request.swift](https://github.com/open-swift/S4/blob/master/Sources/Request.swift)
- [Response.swift](https://github.com/open-swift/S4/blob/master/Sources/Response.swift)

Notice that they are simple data types. Almost like POJOs, if we were coding in JAVA. I totally agree about SocketRequest and Socket with it buffer being an issue. Which opens up another topic- I don't feel like SocketRequest and Socket with its buffer belongs in Request to begin with. That might be another topic of refactoring.

**Comment by vadimeisenbergibm** [comment]
@rfdickerson 

Yep, these are simple data types that should be structs according to the Apple recommendations. Having said that, there is still an issue of performance. When you have a large chunk of data to return in body, and its chunk is copied each time you pass your response between handlers, wouldn't it hurt performance? What happens with value types when they are passed as parameters to functions and returned - are they copied?

> they are simple data types. Almost like POJOs. 

According to [wikipedia](https://en.wikipedia.org/wiki/Plain_Old_Java_Object)
> a Java object which does not follow any of the major Java object models, conventions, or frameworks 

POJO does not mean "simple data type", it can be arbitrarily complex.

> I don't feel like SocketRequest and Socket with its buffer belongs in RouterRequest

If you want to redesign Kitura, I would recommend you to create a class diagram to show what is your proposed class hierarchy. I would also recommend to create two mockup implementations with structs with some significant body size and a dozen of mockup handlers. In one implementation I would pass & return structs by value, in the other implementation - by using inout. Then compare running times. If we see that your design is better and there is no significant performance hit by passing requests with body by value - let's consider rewriting Kitura.

**Comment by rfdickerson** [comment]
Challenge... accepted! :) IBM-Swift/SwiftRuntime#58

**Comment by vadimeisenbergibm** [comment]
@rfdickerson This issue is exactly what I meant, I did not mean it as a challenge :)
Once you have the results, whatever they are, you can publish a blog post about it, or comment on other posts about value types, so people will have more information regarding performance penalty (little or big) of large size value types.

**Comment by vadimeisenbergibm** [comment]
Reply on raywenderlich.com to my comment - https://forums.raywenderlich.com/t/reference-vs-value-types-in-swift-part-1-2/6295/7

**Comment by rfdickerson** [comment]
Good reply at Ray's blog. Btw, [Response](https://github.com/Alamofire/Alamofire/blob/master/Source/Response.swift) in Alamofire is a struct.

**Comment by vadimeisenbergibm** [comment]
[Response in Alamofire](https://github.com/Alamofire/Alamofire/blob/master/Source/Response.swift) does not follow Apple recommendations for copying references. It holds `NSURLRequest `and `NSData` by reference, and does not copy them in init(). `NSMutabaleURLRequest` and `NSMutableData` can be passed to `init()` and then changed from outside. 

So what should we learn from this example? You can write a popular framework while not following Apple recommendations. Whom should we listen to - Alamofire or Apple? I personally prefer Apple. 

Since I think the correct way is to follow the recommendations of Apple, and Alamofire's code does not follow them, for me it is an example of incorrect design and it creates bad impression for Alamofire in my eyes.

 This is my expectation when somebody uses a struct (from [Apple's Swift Programming Language Book](https://developer.apple.com/library/mac/documentation/Swift/Conceptual/Swift_Programming_Language/ClassesAndStructures.html)):

> Any properties stored by the structure are themselves value types, which would also be expected to be copied rather than referenced.

There are no such expectations when somebody uses `class`, so no expectation mismatch would happen if Alamofire would use `class` for their [Response](https://github.com/Alamofire/Alamofire/blob/master/Source/Response.swift).

**Comment by kweinmeister** [comment]
@rfdickerson: Did we reach a conclusion yet? I'm thinking that we consider creating an epic out of this. For any classes where we have unanimous agreement, we create an issue belonging to the epic. For those classes where there are still different points of view, we schedule a meeting to make decisions and move on.

**Comment by rfdickerson** [comment]
I think a lot has been refactored. We can close this issue, I think.

**Comment by kweinmeister** [comment]
Closing per @rfdickerson's suggestion
